General cleanup

klanclos opened the pull request:
Here's a bunch of general cleanup. No functional changes here, but I'd hesitate to call it purely cosmetic, especially when it comes to exception handling. I think the net result is improved readability.

klanclos committed:
Explicitly invoke python3 instead of assuming 'python' is Python 3.x.

klanclos committed:
Remove trailing whitespace.

klanclos committed:
Fix the hashbang line.

klanclos committed:
Normalize the handling of imported modules.

Moved the exception handling related to firewall authentication outside the
two methods involved.

klanclos committed:
Rename some of the attributes and methods related to firewall handling.

klanclos committed:
More comments about the command-line options for ssh.

klanclos committed:
Be more explicit in some of the conditions.

klanclos committed:
Cosmetic fixes to whitespace and conditions.

klanclos committed:
Rearrange the logging setup so it doesn't have to be monkey-patched in
the main context.

klanclos committed:
Tidy up a bunch of conditions.

klanclos committed:
Revise exception handling to log full tracebacks to the debug log.

joshjriley commented on keck_vnc_launcher.py:
```diff
@@ -123,10 +125,16 @@ def start(self):
+            self.firewall_pass = getpass.getpass(f"Password for firewall authentication: ")
+            try:
+                self.is_authenticated = self.authenticate(self.firewall_pass)
+            except:
```
Take a look at handle_fatal_error which was designed to handle errors in a common way, including the stack trace dump to log

klanclos replied:
These are not being handled as fatal errors: they're caught, logged, then the program continues on.

joshjriley reviewed (changes requested):
looks good.  One request to consider using existing handle_fatal_error().

joshjriley reviewed (commented):
I imagine aligning equals signs and other multi-line alignments are not PEP8 compliant and cause other small edit issues, but I absolute LOVE doing this when it makes code more readable at a glance.  In some cases i just do it out of habit unnecessarily.  Anyway, fine with changes but i will agree to disagree.  :)

joshjriley commented on keck_vnc_launcher.py:
```diff
@@ -464,15 +469,24 @@ def get_sessions_requested(self, args):
-        if args.analysis2 is True: sessions.append('analysis2')
-        if args.telanalys is True: sessions.append('telanalys')
-        if args.status    is True: sessions.append('status')
+        if args.control0 is True:
```
Good example of where I prefer readability to non-compliance.

klanclos replied:
In general I find statements in-line with conditions to be less readable than having them broken up. I'm happy to let the PEP decide, but that's hardly generous on my part since it sounds like I agree with it!

joshwalawender replied:
LOL.  I'm half way in between you two. :) I'm ok with a single line having `if condition: do_thing`, but I'm against the column aligning of it.

klanclos replied:
I went and looked, because I didn't know for sure, and PEP 8 does discourage the use of compound statements of this type, it's a soft statement. I'd even go so far as to say it waffles a bit, saying it'd "rather not" have statements like 'if foo == 'blah': do_blah_thing()', but then it says "sometimes it's okay to put an if/for/while with a small body on the same line."

The extra whitespace to align does need to go. I used to do that, and then I realized it interferes with your ability to search through the code. If you still want the single-line compound statements I can put them back on the agree-to-disagree basis.

klanclos replied:
I made this moot by addressing the "todo" in the comment above, iterating through SESSION_NAMES programatically instead of laying it all out explicitly.

joshjriley replied:
excellent thx, i didn't like these being hard coded down in the guts of the code.

joshjriley commented:
I've reviewed most of these and they seem reasonable.  however, I am getting an error on vnc launch a popup with bad argument "-geometry=".  I suspect this might have to do with the one of the numerous changes to be more explicit about checking for None value.  I'll have to debug later or tomorrow.

klanclos committed:
Renamed is_ssh_key_valid to ssh_key_valid.

klanclos committed:
Rename is_ssh_key_valid to ssh_key_valid.

joshwalawender reviewed (commented):
I have a real big problem with changing the imports.  Having a `from x import y` format saves a lot of extraneous characters on lines of code and improves readability (if used properly of course).  What's the point of having calls to `getpass.getpass` instead of `getpass`?  No one is going to be confused as to where is `getpass` coming from?  Ditto for `Thread` or `Telnet` and others.

Other changes here I'm generally fine with.

joshjriley commented on keck_vnc_launcher.py:
```diff
@@ -618,13 +632,13 @@ def launch_vncviewer(self, vncserver, port, geometry=None):
             cmd.append('-ViewOnly')
         #todo: make this config on/off so it doesn't break things
-        if geometry:
+        if geometry is not None:
```
This causes crash because geometry='' coming in.

klanclos replied:
I modified the check to include looking for the empty string, but it seems like the caller should provide None instead of '' if they aren't specifying the geometry. That would be more consistent with how the function is defined. Still, it doesn't hurt to make the condition more explicit.

klanclos replied:
Should now be fixed.

I'm fine with the idea of testing a few different use cases; what's in this branch all works for me, but I can't prove to you which conditions I'm exercising.

joshjriley replied:
I'll help look at them all with a closer eye sometime tomorrow.  We don't have a test suite to test all of these conditionals so there is a small risk there could be more like this.

joshjriley reviewed (changes requested):
One bug found that needs fixing.  I didn't notice it on initial scan thru so I'm a little nervous about the large number of changes to explicit conditions that might have a similar issue.  Each should be looked at and traced to see if assumption is correct (ie could None or '' be passed in and both were assumed to evaluate to as false).

klanclos commented:
It's more typing to write things like `time.sleep(5)` instead of `sleep(5)` but with the former you always know where the function came from. When you're in a namespace and weird global references come in you always have to take a step back, figure out where that reference came from (is it a local function? Is it... something else?) and back-track from there.

There's no ambiguity if you leave the module name intact. Here's a stack exchange question that goes into a bit more depth:

https://softwareengineering.stackexchange.com/questions/187403/import-module-vs-from-module-import-function

Commit:
Also check that the geometry is not the empty string.

Commit:
Programatically iterate over the list of session names instead of
enumerating them explicitly.

klanclos committed:
Rename the variables used in the session getattr() loop.

klanclos committed:
Reserve the all-caps variable name for a global tuple of session names.

klanclos committed:
Remove unused function launch_xterm().

klanclos committed:
Make sure the re-use of existing tunnels defines the vncserver as 'localhost'.

Remove redundant log lines from the menu option processing.

klanclos committed:
Tidy up generating the menu.

klanclos committed:
Close off stdin and stderr from both the ssh tunnels and vncviewer commands.
Closing stdin prevents any of the subprocesses from interfering with menu
input, closing stdout and stderr just gets rid of chatter.

klanclos commented:
I think I'm done with general cleanup for now. That last commit about closing stdin has implications for the menu handling; some of the other cleanup I did was a result of trying to track down what was going wrong there. That means there are now real bugs addressed in here in addition to the general cleanup, sorry about that.

joshwalawender commented:
@klanclos At this point I'm unconvinced about the imports.  I don't see any of the issues mentioned in that thread rising to a significant level here: e.g. we don't use reload(), name collisions seem unlikely, there's really no mystery about where things came from (and at worst it is a scroll to the top of the page).  I'm willing to be convinced, but I'm a long way from it.  It is a style choice and you prefer a different choice than we made, but I see no reason to change this at the last minute before we try to release and it's a bit unfair for you to push that in among other changes.

A lot of the other cleanup looks good, but with the imports folded in and with the issue @joshjriley raised about some of the tests, I don't see how we can merge this without a lot of work.  Can this be separated in to separate PRs?

klanclos commented:
I don't see why we shouldn't follow best practices. This isn't about making it easy on the person that wrote the code, this is about making it easy on someone that has to read it later. The more explicit we can be about what's going on, that's what we should be doing-- while maintaining readability. Sometimes that means being more verbose.

You will never be able to prove to yourself that any one change has been fully exercised unless you set up a unit test suite-- and even then it's only as good as your test coverage. If the changes work for us internally that's all we have to go on. That's going to be true for all pull requests.

joshwalawender commented:
I don't see this as best practices.  I know there are different opinions on that, but it is common practice in python to shorten calls by using the imports in the way we have.  Josh and I have worked on this code for a while, please respect some of the choices we've made.  I'm happy to fix bugs and other minor style issues, but this is more severe.  I'm really happy to have you working on this code during this **extremely** stressful time, but this back and forth is blocking my ability to get actually important fixes in to the code.  Please restore the imports to the original style.

joshjriley commented on keck_vnc_launcher.py:
```diff
@@ -618,13 +632,13 @@ def launch_vncviewer(self, vncserver, port, geometry=None):
             vncargs = vncargs.split()
             cmd = cmd + vncargs
-        if self.args.viewonly:
+        if self.args.viewonly is not None:
```
This caused ViewOnly to always activate because self.args.viewonly is coming in as False by default, not None.

klanclos replied:
Pushed the fix to this branch.

joshjriley reviewed (changes requested):
I reviewed all of the conditionals in detail.  Found one more that needs fixing.  Added comment marking the bug.  Fairly confident that we should be good now on these now.

klanclos committed:
self.args.viewonly is always a boolean, not None.

klanclos commented:
Josh asked me for suggestions on what might be changed/improved, and the way the imports are handled was one of the things that could be changed. That's why I did it. This isn't a question of preference or style, there are good reasons with respect to maintainability to leave your namespace intact.

That said, I'll go ahead and put it back the way it was-- with the exception of sleep().

klanclos committed:
Put back most of the `from x import y` imports.
